File: session-02-dependency-injection/go/order/postgres_repository.go
package order

// Exercise 3: Swap implementation. Change one line in main.go to use this.
type PostgresOrderRepository struct {
	orders []Order
}

func NewPostgresOrderRepository() *PostgresOrderRepository {
	return &PostgresOrderRepository{
		orders: []Order{
			{ID: 1, Drink: "Flat White", Status: "ready"},
			{ID: 2, Drink: "Mocha", Status: "preparing"},
			{ID: 3, Drink: "Cold Brew", Status: "pending"},
			{ID: 4, Drink: "Matcha Latte", Status: "ready"},
		},
	}
}

func (r *PostgresOrderRepository) FindAll() []Order {
	return r.orders
}

func (r *PostgresOrderRepository) FindByID(id int) *Order {
	for _, o := range r.orders {
		if o.ID == id {
			return &o
		}
	}
	return nil
}

File: session-02-dependency-injection/python/order/postgres_repository.py
from order.model import Order
from order.repository import OrderRepository


# Exercise 3: A second repository implementation returning different data.
# Swap get_order_repository() in main.py to use this instead.
class PostgresOrderRepository(OrderRepository):

    def __init__(self):
        self._orders = [
            Order(id=1, drink="Flat White", status="ready"),
            Order(id=2, drink="Mocha", status="preparing"),
            Order(id=3, drink="Cold Brew", status="pending"),
            Order(id=4, drink="Matcha Latte", status="ready"),
        ]

    def find_all(self) -> list[Order]:
        return self._orders

    def find_by_id(self, order_id: int) -> Order | None:
        return next((o for o in self._orders if o.id == order_id), None)
